src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/stores/ResourceFormStore.js

```diff
@@ -87,15 +87,114 @@ export default class ResourceFormStore extends AbstractFormStore implements Form
+        remoteSchema: Schema,
+        localData: Object,
+        remoteData: Object
+    ) {
+        let result = {};
+        if (!localSchema || !remoteSchema) {
+            return result;
+        }
+
+        for (const name in remoteSchema) {
+            const {items: remoteItems, type: remoteType, types: remoteTypes} = remoteSchema[name];
+            const {items: localItems, type: localType, types: localTypes} = localSchema[name] || {};
+
+            if (remoteType === SECTION_TYPE && remoteItems &&
+                localType === SECTION_TYPE && localItems) {
+                result = this.mergeData(
+                    localItems,
+                    remoteItems,
+                    localData,
+                    remoteData
+                );
+                continue;
+            }
+            if (remoteTypes && localTypes
+                && Object.keys(remoteTypes).length > 0 && Object.keys(localTypes).length > 0
+                && localData[name] && remoteData[name]
+                && isArrayLike(localData[name]) && isArrayLike(remoteData[name])
+            ) {
+                for (const key of remoteData[name].keys()) {
+                    const remoteChildData = remoteData[name][key];
+                    const localChildData = toJS(localData[name].length > key ? localData[name][key] || {} : {});
+
+                    const localChildSchema = localTypes[localChildData.type]?.form;
+                    const remoteChildSchema = remoteTypes[remoteChildData.type].form;
+
+                    if (!(localChildData.type in remoteTypes)) {
+                        //set default type
+                        localChildData.type = remoteSchema[name].defaultType;
+                    }
+
+                    const resultChildData = this.mergeData(
+                        localChildSchema,
+                        remoteChildSchema,
+                        localChildData,
+                        remoteChildData
+                    );
+
+                    if (!result[name]) {
+                        result[name] = [];
+                    }
+
+                    if (Object.keys(resultChildData).length === 0) {
+                        if (localChildData.type === remoteChildData.type){
+                            result[name][key] = localChildData;
+                        } else {
+                            result[name][key] = remoteChildData;
+                        }
+                    } else {
+                        result[name][key] = resultChildData;
+                    }
+                }
+
+                continue;
+            }
+
+            if (!localData[name]) {
+                result[name] = remoteData[name];
+                if (remoteData.type) {
+                    result.type = remoteData.type;
+                }
+
+                if (remoteData.settings) {
+                    result.settings = localData?.settings || remoteData.settings;
+                }
```

Comment:
think i would move this into the `block` case above to make clear that we need to do this only for blocks 🙈